WebDriver Bidi tests with new Py3 WebSockets dependencies

k7z45 opened the pull request:
Created a prototype involving WebDriver Bidi client according to [1].

The dependencies for the WebSockets lib[2] is added in third party lib. Added the dependency for pytest-asyncio[3] in this PR as well.

Later versions of pytest[4] and pluggy[5] are needed for pytest-asyncio and they are included here. Pluggy is also needed and it's already included from rebase.

The change includes:

A simple WebDriver Bidi capabilities test to demonstrate enabling WebSockets according to the protocol.
Upgraded pytest and pluggy third party dependencies
Added an async style test that enables WebSocket by directly specifying the corresponding capability.
Add BidiSession class in WebDriver client for use in fixture to conveniently create a Bidi Session and tests to demonstrate behavior between tests.
[1] #26015 (comment)
[2] https://github.com/aaugustin/websockets
[3] https://github.com/pytest-dev/pytest-asyncio
[4] https://github.com/pytest-dev/pytest
[5] https://github.com/pytest-dev/pluggy

k7z45 committed:
Add a test for WebDriver capabilities for webSocketUrl

Test establishing a Bidirectional Session capability support

k7z45 committed:
Add an async style test for connecting to WebDriver

Test connecting to WebDriver session using websockets
and send a message in async/await style

k7z45 committed:
Add BidiSession class and tests using modified session fixture

1. Added a BidiSession class in bidi.py under tools/webdriver/webdriver
that will preset capability for websocket connection.
Created async versions of start and end for the class to handle
establishing and closing websocket connection.
2. Modify session fixture to take an extra boolean argument "bidi"
for creating bidi session. Change the fixture to async and use
await when dealing with bidi session start/end.
3. Set event loop scope to session(assuming it would reuse the same
event loop for all tests) to avoid event loop is closed error when
closing websocket connection.
4. Added tests for different bidi/classic session creation scenario
and tested connect/send/close.

k7z45 commented:
To run the test, I believe still need to do
`pip install pytest-asyncio`
locally.
Then
`./wpt --py3  run --log-mach=- --channel=dev chrome webdriver/tests/new_session/websocket_url.py `
and
`./wpt --py3  run --log-mach=- --channel=dev chrome webdriver/tests/websockets/connect.py   `

stephenmcgruer reviewed (commented):
Hi Shengfa, thanks for the PR! I've started to take a look at it, but a quick question:

> The actual dependencies for the WebSockets lib[2] [...] are not included in this PR. They are installed locally using pip install.

Is this comment (from the description) still accurate? It looks like you vendored `websockets` into `third_party` and added it to `localpaths.py` so I wasn't sure.

k7z45 commented:
> Hi Shengfa, thanks for the PR! I've started to take a look at it, but a quick question:
> 
> > The actual dependencies for the WebSockets lib[2] [...] are not included in this PR. They are installed locally using pip install.
> 
> Is this comment (from the description) still accurate? It looks like you vendored `websockets` into `third_party` and added it to `localpaths.py` so I wasn't sure.

Thanks for looking at it. That's not true anymore. I will modify the description tomorrow.

k7z45 commented on tools/webdriver/webdriver/client.py:
```diff
@@ -753,6 +760,60 @@ def execute_async_script(self, script, args=None):
+        if caps:
+            caps.setdefault("alwaysMatch", {}).update({"webSocketUrl": True})
+        else:
+            caps = {"alwaysMatch": {"webSocketUrl": True} }
```
Fixed.

k7z45 commented on tools/webdriver/webdriver/client.py:
```diff
@@ -1,3 +1,7 @@
+import copy
+import websockets
+import asyncio
```
Removed. Thought it's needed for async def functions.

k7z45 commented on webdriver/tests/support/fixtures.py:
```diff
@@ -127,19 +147,35 @@ def session(capabilities, configuration, request):
 
-    # If there is a session with different capabilities active, end it now
+    is_cur_bidi = type(_current_session) == webdriver.BidiSession
+    # If there is a session with different capabilities active or the current session 
```
Fixed.

jgraham commented:
Is it possible to clean up the commits here to make reviewing eadier? In particular it would be great to have a single initial commit that vendors in third party code, and then one or more subsequent commits with the additional changes.

k7z45 commented:
> Is it possible to clean up the commits here to make reviewing easier? In particular it would be great to have a single initial commit that vendors in third party code, and then one or more subsequent commits with the additional changes.

Hi James,
I tried to to git rebase -i HEAD~16
with
pick 4b64d94ed9 Upgrade pytest to 6.1.1 to be used with pytest-asyncio
squash ff59ff2c63 Remove tools/third_party/pytest for subtree add
squash bcc387416d Squashed 'tools/third_party/pytest/' content from commit 0ad20b533f
squash 0b4aec05b1 Squashed 'tools/third_party/websockets/' content from commit 139085fe26
squash 182e94dc2e Add websockets to localpaths
squash fbedffb6a5 Squashed 'tools/third_party/iniconfig/' content from commit 4e20f6bcde
squash 0e7f71faf0 Add iniconfig in localpaths,
unfortunately, I don't know how to resolve all the merge conflicts correctly.

stephenmcgruer commented:
> unfortunately, I don't know how to resolve all the merge conflicts correctly.

Trying to rebase a set of changes which include a `subtree add` and a merge (`89e15e08325f8a8c4ce062477ecb8e5f6d49cdeb Merge branch 'master' into websockets`) is difficult. It may be easier building the branch from scratch via cherry-picks (and re-doing the `subtree add`). Let me know if you need some assistance doing that.

k7z45 commented:
Hi James,
Could you clarify what commits you expect to see, please?
I am having trouble trying to squash the `subtree add` commits into one commit.

jgraham commented:
Baiscally I want all the commits that vendor in libraries first and then either a single commit with your changes or a number of logically seperate commits with your changes, with no merge commits except those from subtree. The ise case is to be able to filter out all the vendored code when doing the review.

k7z45 commented:
Got it, thanks for the clarification.

k7z45 commented:
Re-applied commits with subtree add, cherry-pick and rebase so that they are better organized for review.

stephenmcgruer commented on tools/third_party/pytest/src/_pytest/assertion/rewrite.py:
```diff
@@ -27,7 +27,7 @@
 
 from _pytest._io.saferepr import saferepr
-from _pytest._version import version
+#from _pytest._version import version
```
I wonder if these problems are because our `.tox` version of pytest would now be out of sync with the third_party version.

Could you try updating the pytest versions in `tools/wpt/tox.ini`, `tools/wptrunner/tox.ini`, `tools/wave/tox.ini`, and `tools/tox.ini` to be the same as the one you vendored in?

k7z45 replied:
After changing the `rewrite.py ` back and `tools/wpt/tox.ini`, `tools/wptrunner/tox.ini`, `tools/wave/tox.ini`, and `tools/tox.ini` versions, run 
`python3 wpt  run --log-mach=- --channel=dev chrome webdriver/tests/websockets/connect.py`.
I still got
` 0:07.60 INFO STDERR:   File "/usr/local/google/home/shengfa/Workspace/wpt/tools/third_party/pytest/src/_pytest/assertion/rewrite.py", line 30, in <module>
 0:07.60 INFO STDERR:     
 0:07.60 INFO STDERR: from _pytest._version import version
 0:07.60 INFO STDERR: ModuleNotFoundError
 0:07.60 INFO STDERR: : 
 0:07.60 INFO STDERR: No module named '_pytest._version'
`

k7z45 replied:
There is an import error from `tools/third_party/pytest/src/_pytest/__init__.py` to import `from ._version import version as __version__` but `rewrite.py` is still trying to use the value directly.

k7z45 replied:
I think it's probably some path or install issue. In `tools/third_party/pytest/setup.cfg` it has
`[check-manifest]
ignore =
    src/_pytest/_version.py
`
so I think that file should somehow be generated for the import to succeed but I am not sure how to achieve that.

stephenmcgruer replied:
Ok, so yes this is because `pytest` now uses a setup that builds the `_version.py` file. We should probably be grabbing the built pytest from pip directly, e.g. https://pypi.org/project/pytest/6.1.1/#files

I tried downloading the `.tar.gz`, unzipping it, and copying it into our vendored `pytest` directory. I think executed `git add -f tools/third_party/pytest` and got:

```
$ git status
HEAD detached from shengfa/websockets
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        deleted:    tools/third_party/pytest/.coveragerc
        deleted:    tools/third_party/pytest/.gitattributes
        deleted:    tools/third_party/pytest/.gitblameignore
        deleted:    tools/third_party/pytest/.github/FUNDING.yml
        deleted:    tools/third_party/pytest/.github/ISSUE_TEMPLATE/1_bug_report.md
        deleted:    tools/third_party/pytest/.github/ISSUE_TEMPLATE/2_feature_request.md
        deleted:    tools/third_party/pytest/.github/ISSUE_TEMPLATE/config.yml
        deleted:    tools/third_party/pytest/.github/PULL_REQUEST_TEMPLATE.md
        deleted:    tools/third_party/pytest/.github/config.yml
        deleted:    tools/third_party/pytest/.github/labels.toml
        deleted:    tools/third_party/pytest/.github/workflows/main.yml
        deleted:    tools/third_party/pytest/.github/workflows/release-on-comment.yml
        deleted:    tools/third_party/pytest/.gitignore
        deleted:    tools/third_party/pytest/.pre-commit-config.yaml
        deleted:    tools/third_party/pytest/.readthedocs.yml
        deleted:    tools/third_party/pytest/.travis.yml
        new file:   tools/third_party/pytest/PKG-INFO
        modified:   tools/third_party/pytest/setup.cfg
        new file:   tools/third_party/pytest/src/_pytest/_version.py
        new file:   tools/third_party/pytest/src/pytest.egg-info/PKG-INFO
        new file:   tools/third_party/pytest/src/pytest.egg-info/SOURCES.txt
        new file:   tools/third_party/pytest/src/pytest.egg-info/dependency_links.txt
        new file:   tools/third_party/pytest/src/pytest.egg-info/entry_points.txt
        new file:   tools/third_party/pytest/src/pytest.egg-info/not-zip-safe
        new file:   tools/third_party/pytest/src/pytest.egg-info/requires.txt
        new file:   tools/third_party/pytest/src/pytest.egg-info/top_level.txt
```

That seems like a sensible diff to me.

However when I `wpt run` I get `iniconfig` errors from `findpaths.py` in pytest - did you run into that?

```
 0:08.73 INFO STDERR: Exception in thread 
 0:08.73 INFO STDERR: Thread-1
 0:08.73 INFO STDERR: :
 0:08.73 INFO STDERR: Traceback (most recent call last):
 0:08.73 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/wptrunner/wptrunner/executors/base.py", line 668, in _run
 0:08.73 INFO STDERR:     
 0:08.74 INFO STDERR: self.result = True, self.func(self.session, self.path, self.timeout)
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/wptrunner/wptrunner/executors/base.py", line 630, in do_wdspec
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: return pytestrunner.run(path,
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/wptrunner/wptrunner/executors/pytestrunner/runner.py", line 42, in run
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: do_delayed_imports()
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/wptrunner/wptrunner/executors/pytestrunner/runner.py", line 24, in do_delayed_imports
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: import pytest
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/third_party/pytest/src/pytest/__init__.py", line 5, in <module>
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: from _pytest.assertion import register_assert_rewrite
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/third_party/pytest/src/_pytest/assertion/__init__.py", line 8, in <module>
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: from _pytest.assertion import rewrite
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/third_party/pytest/src/_pytest/assertion/rewrite.py", line 37, in <module>
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: from _pytest.config import Config
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/third_party/pytest/src/_pytest/config/__init__.py", line 42, in <module>
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: from .findpaths import determine_setup
 0:08.74 INFO STDERR:   File "/home/smcgruer/github/wpt/tools/third_party/pytest/src/_pytest/config/findpaths.py", line 10, in <module>
 0:08.74 INFO STDERR:     
 0:08.74 INFO STDERR: import iniconfig
 0:08.74 INFO STDERR: ModuleNotFoundError
 0:08.74 INFO STDERR: : 
 0:08.74 INFO STDERR: No module named 'iniconfig'
```

k7z45 replied:
Thanks! Updated pytest with
> I tried downloading the .tar.gz, unzipping it, and copying it into our vendored pytest directory. I think executed git add -f tools/third_party/pytest

and no longer need to change version manually.

> However when I wpt run I get iniconfig errors from findpaths.py in pytest - did you run into that?

This should not happen as iniconfig is added in tools/third_party and localpath in one of the commits.

stephenmcgruer reviewed (commented):
This is looking good, but I wanted to confirm:

1. `websockets` was added to localpath, but does that work for `tools/webdriver/webdriver/client.py`? I thought we still needed to solve the problem that previously `tools/webdriver/*` didn't have any vendored dependencies and this would be the first one...
2. What's the plan for `pytest-asyncio` - are we going to vendor that as well?

k7z45 commented:
> This is looking good, but I wanted to confirm:
> 
> 1. `websockets` was added to localpath, but does that work for `tools/webdriver/webdriver/client.py`? I thought we still needed to solve the problem that previously `tools/webdriver/*` didn't have any vendored dependencies and this would be the first one...
Adding websockets to third party and adding it in localpath seems to be working. The `connect.py` test can run and would fail if I remove `tools/third_party/websockets` with errors such as
`>       self.websocket_transport = await websockets.connect(self.capabilities["webSocketUrl"])
E       AttributeError: module 'websockets' has no attribute 'connect'
`
> 2. What's the plan for `pytest-asyncio` - are we going to vendor that as well?
I tried vendor in the dependency with subtree add and adding to localpath but got error such as
` 0:09.82 INFO STDOUT: ________________________________________ ERROR collecting webdriver/tests/websockets/connect.py ________________________________________
 0:09.82 INFO STDOUT: 'asyncio' not found in `markers` configuration option`

stephenmcgruer commented:
Ok, we're making progress I think!

I downloaded pytest-asyncio from https://pypi.org/project/pytest-asyncio/#files, un-tarred it, added it to `third_party` and to `localpaths.py`, and that seemed to work. I am now able to successfully run:

`./wpt run --log-mach=- --channel=dev chrome webdriver/tests/websockets/connect.py`

@k7z45 - was there a reason you didn't add pytest-asyncio in this PR?

k7z45 committed:
Add pytest-asyncio to localpaths

k7z45 committed:
Add pytest-asyncio to tools third_party

k7z45 commented:
> Ok, we're making progress I think!
> 
> I downloaded pytest-asyncio from https://pypi.org/project/pytest-asyncio/#files, un-tarred it, added it to `third_party` and to `localpaths.py`, and that seemed to work. I am now able to successfully run:
> 
> `./wpt run --log-mach=- --channel=dev chrome webdriver/tests/websockets/connect.py`
> 
> @k7z45 - was there a reason you didn't add pytest-asyncio in this PR?

Thanks!
That works. I have tried un-tarring it before but could not get it working. Must have done something wrong.
I have updated the commits to include it.

k7z45 committed:
Add websockets and iniconfig to tools/localpaths.py

k7z45 committed:
Add pytest from pip directly

Use git add -f tools/third_party/pytest to include version file.
This includes _version.py and pytest.egg-info folder.

stephenmcgruer commented:
So we can see from https://community-tc.services.mozilla.com/tasks/Gm3vHxznSVWScWRjyx03uQ/runs/0/logs/public/logs/live.log that the websockets library is not being found. This is probably because that flow doesn't go through `wpt run` so doesn't import localpaths.py.

Se we need to decide how to deal with tools/webdriver/ now having dependencies.

@jgraham - thoughts? I'm tempted to just have `tools/webdriver/` code pull in `localpaths` but perhaps we don't want to expand use of that file :D

jgraham commented:
I think we should make that test frontend import localpaths. In general I think it makes sense to write the library code itself on the assumption that something™ makes the dependencies available and then have the frontend choose how to do that. Ideally we'd have less localpaths.py in our lives…

jgraham commented on webdriver/tests/websockets/connect.py:
```diff
@@ -0,0 +1,12 @@
+async def test_websocket_url_connect(session):
+    websocket_url = session.capabilities["webSocketUrl"]
+    async with websockets.connect(websocket_url) as websocket:
+        await websocket.send("Hello world!")
```
The idea here is that this can end up returning an error on the socket, but we don't care?

Anyway it seems reasonable. I don't think the commit message should mention Chrome though; this is surely for any bidi implementation.

k7z45 replied:
Yes, that's the idea.
I will replace ChromeDriver with WebDriver.

jgraham commented on tools/webdriver/webdriver/client.py:
```diff
@@ -752,6 +758,60 @@ def execute_async_script(self, script, args=None):
+        capabilities = self.__enable_websocket(capabilities)
+        super().__init__(host, port, url_prefix, capabilities, extension)
+
+    def __enable_websocket(self, caps):
```
I think we genrally stick to single underscore names to avoid the name mangling behaviour.

k7z45 replied:
Done.

jgraham commented on tools/webdriver/webdriver/client.py:
```diff
@@ -752,6 +758,60 @@ def execute_async_script(self, script, args=None):
+        :return: Dictionary with `capabilities` and `sessionId`.
+        """
+        value = super().start()
+        # Refresh websocket connection for every test to have a clean
```
I don't think this library should contain this kind of test-specific detail, but it should go into `conftest.py`. Closing the websocket connection on every test seems pretty dubious; there isn't much state tied to the connection itself rather than the session, so I'd initially not do this and see if we are running into issues.

k7z45 replied:
Removed the close and re-connect logic.

jgraham commented on tools/webdriver/webdriver/client.py:
```diff
@@ -752,6 +758,60 @@ def execute_async_script(self, script, args=None):
     def screenshot(self):
         return self.send_session_command("GET", "screenshot")
 
+class BidiSession(Session):
```
nit: I think there would be something to be said for this being in a different file.

k7z45 replied:
Done.

jgraham commented on webdriver/tests/support/fixtures.py:
```diff
@@ -38,6 +41,20 @@ def pytest_generate_tests(metafunc):
+            metafunc.parametrize("bidi", marker.args, ids=None)
+
+# Reuse one event loop for all tests in the test session
+# so that we won't have the event loop closed issue when
```
What's the "event loop closed issue"? (I assume it's soemthing like the event loop shutting down when there aren't any registered users, but it would be good to be clear).

k7z45 replied:
When I don't specify this, I get error such as
```
coro_or_future = <Task cancelled name='Task-10' coro=<WebSocketCommonProtocol.transfer_data() done, defined at /usr/local/google/home/shengfa/Workspace/wpt2/tools/third_party/websockets/src/websockets/protocol.py:818>>

    def ensure_future(coro_or_future, *, loop=None):
        """Wrap a coroutine or an awaitable in a future.

        If the argument is a Future, it is returned directly.
        """
        if coroutines.iscoroutine(coro_or_future):
            if loop is None:
                loop = events.get_event_loop()
            task = loop.create_task(coro_or_future)
            if task._source_traceback:
                del task._source_traceback[-1]
            return task
        elif futures.isfuture(coro_or_future):
            if loop is not None and loop is not futures._get_loop(coro_or_future):
>               raise ValueError('The future belongs to a different loop than '
                                 'the one specified as the loop argument')
E               ValueError: The future belongs to a different loop than the one specified as the loop argument
```
I found similar issue here: https://github.com/pytest-dev/pytest-asyncio/issues/171
I think it's because when I don't specify the scope to be session, it's by default function. Then the event loop will be closed after one test and restarted; hence the error "belongs to a different loop".

jgraham replied:
That makes sense, thanks. Could you update the comment to explain this. Something like
```py
# Ensure that the event loop is restarted once per session rather than the default  of once per test
# if we don't do this, tests will try to reuse a closed event loop and fail with an error that the "future
# belongs to a different loop"
```

k7z45 replied:
Done, thanks for the suggestion!

jgraham commented on webdriver/tests/support/fixtures.py:
```diff
@@ -127,19 +147,35 @@ def session(capabilities, configuration, request):
     caps = {"alwaysMatch": caps}
 
-    # If there is a session with different capabilities active, end it now
+    is_cur_bidi = type(_current_session) == webdriver.BidiSession
```
I might use `isinstance` here.

k7z45 replied:
Done.

jgraham commented on webdriver/tests/support/fixtures.py:
```diff
@@ -127,19 +147,35 @@ def session(capabilities, configuration, request):
     if _current_session is not None and (
-            caps != _current_session.requested_capabilities):
-        _current_session.end()
+            (not _current_session.match(caps)) or
```
Can the capabilities match but the bidi value mismatch?

k7z45 replied:
Yes, user can specify the webSocketUrl capability to true with or without @pytest.mark.bidi(True).
If user specify the webSocketUrl to true without the marker first and then do it again with the same capability but with the marker. The capabilities match but bidi value mismatch. In this case, we still create a new instance of session. Normal use case would be to just use the marker.

stephenmcgruer commented:
> I think we should make that test frontend import localpaths. In general I think it makes sense to write the library code itself on the assumption that something™ makes the dependencies available and then have the frontend choose how to do that. Ideally we'd have less localpaths.py in our lives…

Makes sense to me. @k7z45 - I believe the change needed then is to have `resources/test/conftest.py` import localpaths before it imports webdriver.

k7z45 committed:
Import localpaths in resources/test/conftest.py

k7z45 commented:
Thanks for reviewing the change! PTAL.

jgraham commented on tools/webdriver/webdriver/__init__.py:
```diff
@@ -36,3 +35,5 @@
     UnknownMethodException,
     UnsupportedOperationException,
     WebDriverException)
+from .bidi_session import (
```
Having an `_` in module names is quite unusual in Python. Perhaps we could just have a name like `bidi`?

k7z45 replied:
Done.

jgraham reviewed (commented):
This is really nearly ready to merge. I'm very excited for this to be landed!

stephenmcgruer reviewed (approved):
LGTM, but (a) please wait for jgraham to also approve, and (b) can we rebase the commits before landing so that the review changes are squashed back into the appropriate original commit? (E.g. 102b879478f86bc3da32ef4003b281f6c0b16798 would be squashed into 8a43e9fcd3a2319052c2611a44e33723355ab9af, etc)

k7z45 commented:
> LGTM, but (a) please wait for jgraham to also approve, and (b) can we rebase the commits before landing so that the review changes are squashed back into the appropriate original commit? (E.g. [102b879](https://github.com/web-platform-tests/wpt/commit/102b879478f86bc3da32ef4003b281f6c0b16798) would be squashed into [8a43e9f](https://github.com/web-platform-tests/wpt/commit/8a43e9fcd3a2319052c2611a44e33723355ab9af), etc)

Will do that once approved. Thanks for the reminder!

whimboo commented on webdriver/tests/support/fixtures.py:
```diff
@@ -24,6 +25,8 @@ def pytest_configure(config):
     config.addinivalue_line("markers",
         "capabilities: mark test to use capabilities")
+    config.addinivalue_line("markers",
+        "bidi: mark test to use bidi")
```
Instead of a marker, which would mean that we have to add it before each and every test method, having a dedicated `bidi_session` fixture would cause way lesser code to write:

```
async def test_log_events(bidi_session, url):
    bidi_session.url = url("/tests/bidi/support/log_events.html")
    ...
```

That way we can also factor out all the bidi specific code from the `session` fixture itself, and have a clear separation for the BiDi extension.

What do you think?

k7z45 replied:
I think that's a good idea. It increased some code in the fixture by not reusing the session constructor but would reduce markers as number of bidi tests increases.

I created a new current_bidi_session global variable in the new commit and modified the tests to use the new bidi_session fixture. I think this is a big change compare to previous changes. I would consider reverting the change and merge what we have so far and create a separate issue to handle it in case there is unintended effect.

whimboo replied:
Would be fine with me to move it to a follow-up issue/PR.

k7z45 replied:
Thanks, reverted the latest commit.

whimboo commented on webdriver/tests/support/fixtures.py:
```diff
@@ -38,6 +41,20 @@ def pytest_generate_tests(metafunc):
+    """Change event_loop fixture to session level."""
+    loop = asyncio.get_event_loop_policy().new_event_loop()
+    yield loop
+    loop.close()
```
With the dedicated session approach this can also become part of the`bidi_session` fixture.

k7z45 replied:
I created the bidi_session fixture in the same fixtures.py. Not sure how to make this "become part of the ```bidi_session``` fixture".

whimboo replied:
Lets move this to the new issue/PR. Mind moving over all the remaining open conversations?

k7z45 replied:
Will do.

whimboo commented on webdriver/tests/websockets/connect.py:
```diff
@@ -0,0 +1,43 @@
+import pytest
```
I would suggest to put all the BiDi tests under a separate `bidi/` subfolder, and I think that this should also be done for the websocket tests.

Also the tests in this file are all around new sessions, so maybe a location at `webdriver/tests/bidi/new_session` would make sense?

k7z45 replied:
Thanks for the suggestion! This is done.

whimboo commented on webdriver/tests/websockets/connect.py:
```diff
@@ -0,0 +1,43 @@
+    await session.websocket_transport.send("test_bidi_session_1")
+    await session.websocket_transport.close()
+
+# bidi session following a bidi session with the same capabilities
```
This makes the test dependent on a strict order, and `test_bidi_session_1` being enabled. To get rid of this all the code should be put into a single test.

k7z45 replied:
One of the reasons for the tests is testing the behavior from re-using websocket_transport between tests with/without capability change. I don't think I can achieve that with a single test.

whimboo replied:
We might need a similar `new_session` fixture for BiDi so that you can create multiple new sessions with different capabilities within a single test.

whimboo commented on webdriver/tests/websockets/connect.py:
```diff
@@ -0,0 +1,43 @@
+@pytest.mark.asyncio
+@pytest.mark.bidi(True)
+@pytest.mark.capabilities({"acceptInsecureCerts": True})
+async def test_bidi_session_3(session):
```
Maybe you can find a better name for the suffix like `test_bidi_session_with_additional_caps`?

k7z45 replied:
I changed the tests with better namings.

jgraham commented on webdriver/tests/support/fixtures.py:
```diff
@@ -18,16 +18,14 @@
 
 
 _current_session = None
+_current_bidi_session = None
```
I'm worried about having more than one global session. I think it would make sense to have a `bidi_session` fixture but for it to actually share the same session as the `_current_session`; if the bidi status changes we'd need to restart the session just like we do with the capabilities.

k7z45 committed:
Moved bidi test to its own folder

k7z45 commented:
Reverted the latest commit but kept the new bidi test folder. Squashed commits from comments into appropriate original commit.
Will open another issue to address remaining opened conservation for

1. Creating a bidi_session fixture instead of existing bidi marker to reduce user test code setup for the marker.
2. Re-use global current_session instead of creating another global bidi session
3. > We might need a similar `new_session` fixture for BiDi so that you can create multiple new sessions with different capabilities within a single test.

whimboo reviewed (approved):
I'm ok with moving the remaining parts to a new issue / PR. @jgraham mind doing a review pass yourself?

jgraham reviewed (approved):
OK, I'm happy for this to land to get things unblocked, but I agree that the proposed followups will create a nicer API.

jgraham commented:
I think think one needs to be landed by manual push so preserve the merge commits?

stephenmcgruer commented:
> I think think one needs to be landed by manual push to preserve the merge commits?

Yes, I believe so. I will own syncing with Shengfa to get this landed (I'm not sure what permissions we'll need, but I assume I have them and he might not...)

stephenmcgruer commented:
I have manually pushed the two `subtree adds` (which generate merge commits and so cannot be landed via GitHub's UI). We believe the rest of the PR can be landed via GitHub's UI, so shengfa has rebased it and we're waiting for the CI to run.
